dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -81,93 +53,83 @@ public Set<Relation> getRelations() {
      * @return The relation in this model named {@code name}, or {@code null} if no such exists.
      */
     public Relation getRelation(String name) {
-        Relation relation = relationMap.get(name);
-        if (relation == null) {
-            if (RelationNameRepository.contains(name)) {
-                relation = getBasicRelation(name).setName(name);
-                addRelation(relation);
+        for (Relation r : relations) {
+            if (r.hasName(name)) {
+                return r;
             }
         }
-        return relation;
-    }
-
-    public Relation getRelation(Class<?> cls, Object... args) {
-        Class<?>[] argClasses = getArgsForClass(cls);
-        try {
-            Method method = cls.getMethod("makeTerm", argClasses);
-            String term = (String) method.invoke(null, args);
-            Relation relation;
-            if (relationMap.containsKey(term)) {
-                relation = relationMap.get(term);
-            } else {
-                Constructor<?> constructor = cls.getConstructor(argClasses);
-                relation = (Relation) constructor.newInstance(args);
-                addRelation(relation);
-            }
-            return relation;
-        } catch (Exception e) {
-            e.printStackTrace();
+        if (!RelationNameRepository.contains(name)) {
             return null;
```

Comment:
Isn't it dangerous to return `null`? Wouldn't be better that throw an exception if this is not expected?

Reply:
This is intended behaviour if you query for a non-existent relation.
One could return `Optional<Relation>` instead, but I think returning `null` is also fine.

Reply:
I could let it throw and provide a method `hasRelation(String)` for querying.  Most of the callers expect an object and would throw an ambiguous NullPointer later.

Reply:
If you do this, you have to be careful about the semantics of `hasRelation(String)` for base relations that have not been "created" yet (i.e. those that would get created implicitly).

Honestly, I'm not really a fan of the implicit construction of named base relations upon querying to begin with. But I don't know how much code actually relies on this and how hard it is to get rid of that.

Reply:
`getRelation(String)` now throws an InvalidArgument in those cases.